runtime/lib/ttmetal/command_queue.cpp

```diff
@@ -0,0 +1,367 @@
+  std::array<uint32_t, 2> tensorShape;
+  assert(layout->stride()->size() >= 2);
+  tensorShape[1] = layout->stride()->Get(tensorRank - 2);
+  tensorShape[0] =
+      layout->stride()->Get(0) * command->ref()->desc()->shape()->Get(0);
+
+  auto pageShape = shardShape;
+  ShardSpecBuffer shardSpecBuffer(shardSpec, pageShape, tensorShape);
+
+  uint64_t gridVolume = 1;
+  for (auto dim2dRange : *layout->core_range_set()) {
+    gridVolume *= dim2dRange->size().x() * dim2dRange->size().y();
+  }
+
+  assert(layout->memory_desc()->memory_space() ==
+             ::tt::target::MemorySpace::DeviceDRAM ||
+         layout->memory_desc()->memory_space() ==
+             ::tt::target::MemorySpace::DeviceL1);
+  BufferType bufferType = layout->memory_desc()->memory_space() ==
+                                  ::tt::target::MemorySpace::DeviceDRAM
+                              ? BufferType::DRAM
+                              : BufferType::L1;
+  uint64_t size = gridVolume * layout->memory_desc()->size();
+  auto shardedBufferConfig = ShardedBufferConfig{
+      .device = device,
+      .size = size,
+      .page_size = size,
+      .buffer_type = bufferType,
+      .buffer_layout = TensorMemoryLayout::HEIGHT_SHARDED,
+      .shard_parameters = shardSpecBuffer,
+  };
+  std::shared_ptr<::tt::tt_metal::Buffer> buffer =
+      ::tt::tt_metal::CreateBuffer(shardedBufferConfig);
+  buffer->set_address(command->ref()->address());
+  buffers[command->ref()->global_id()] = buffer;
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::DeallocateBufferCommand const *command) {
+  auto iter = buffers.find(command->ref()->global_id());
+  assert(iter != buffers.end() && "Buffer not allocated");
+  assert(iter->second != nullptr && "Buffer already deallocated");
+  ::tt::tt_metal::DeallocateBuffer(*iter->second);
+  iter->second.reset();
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::CreateEventCommand const *command) {
+  assert(events.find(command->ref()->global_id()) == events.end());
+  events[command->ref()->global_id()] =
+      std::make_shared<::tt::tt_metal::Event>();
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::EnqueueRecordEventCommand const *command) {
+  auto event = events.at(command->ref()->global_id());
+  ::tt::tt_metal::EnqueueRecordEvent(*cq, event);
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::EnqueueWaitForEventCommand const *command) {
+  auto event = events.at(command->ref()->global_id());
+  ::tt::tt_metal::EnqueueWaitForEvent(*cq, event);
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::EventSynchronizeCommand const *command) {
+  auto event = events.at(command->ref()->global_id());
+  ::tt::tt_metal::EventSynchronize(event);
+}
+
+void CQExecutor::execute(
+    ::tt::target::metal::EventQueryCommand const *command) {
+  auto event = events.at(command->ref()->global_id());
+  (void)::tt::tt_metal::EventQuery(
+      event); // todo, we need flatbuffer support for tracking and doing
```

Comment:
Don't quite understand how this all works yet, but in my head having hard time visualizing how EventQuery (returns bool in metal, true if event is completed, false if not completed) can be used here or what makes sense... does the bool value need to be propogated up, and does that even make sense?

Reply:
At the moment it makes no sense, in order to support this we need to augment flatbuffer:

```
table BoolRef {
  global_id: uint32;
}

table EventQueryCommand {
  ref: EventRef;
  out: BoolRef;
}
```

And then this code would change to be:
```
boolVars[command->out()->global_id()] = ::tt::tt_metal::EventQuery(event);
```

And then somewhere later in the flatbuffer program it could reference `boolVars[command->out()->global_id()]` and conditionally do something.  But I actually don't have a use case at the moment to leverage this.

In a separate context I could envision a frontend API:
```
bool eventQuery(Event event);
```

For which a frontend runtime might want to query if work has finished.